CommonTools/RecoAlgos/plugins/SimClusterRecHitAssocitionProducer.cc

```diff
@@ -0,0 +1,136 @@
+// system include files
+#include <memory>
+#include <string>
+
+// user include files
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/Framework/interface/ESHandle.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "SimDataFormats/TrackingAnalysis/interface/TrackingParticle.h"
+#include "SimDataFormats/CaloAnalysis/interface/SimClusterFwd.h"
+#include "SimDataFormats/CaloAnalysis/interface/SimCluster.h"
+#include "DataFormats/CaloRecHit/interface/CaloRecHit.h"
+#include "SimDataFormats/CaloHit/interface/PCaloHit.h"
+#include "SimDataFormats/CaloHit/interface/PCaloHitContainer.h"
+
+#include "DataFormats/Common/interface/Association.h"
+
+#include "FWCore/Utilities/interface/transform.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include <set>
+
+//
+// class decleration
+//
+typedef std::pair<size_t, float> IdxAndFraction;
+
+class SimClusterRecHitAssociationProducer : public edm::stream::EDProducer<> {
+public:
+  explicit SimClusterRecHitAssociationProducer(const edm::ParameterSet&);
+  ~SimClusterRecHitAssociationProducer() override;
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+
+  std::vector<edm::InputTag> caloRechitTags_;
+  std::vector<edm::EDGetTokenT<edm::PCaloHitContainer>> caloSimhitCollectionTokens_;
+  //std::vector<edm::EDGetTokenT<std::vector<PSimHit>> trackSimhitCollectionTokens_;
+  std::vector<edm::EDGetTokenT<edm::View<CaloRecHit>>> caloRechitCollectionTokens_;
+  edm::EDGetTokenT<SimClusterCollection> scCollectionToken_;
+};
+
+SimClusterRecHitAssociationProducer::SimClusterRecHitAssociationProducer(const edm::ParameterSet& pset)
+    :  //caloSimhitCollectionTokens_(edm::vector_transform(pset.getParameter<std::vector<edm::InputTag>>("caloSimHits"),
+      //  [this](const edm::InputTag& tag) {return mayConsume<edm::PCaloHitContainer>(tag); })),
+      //trackSimhitCollectionTokens_(edm::vector_transform(pset.getParameter<edm::InputTag>("trackSimHits"),
+      //    [this](const edm::InputTag& tag) {return mayConsume<std::vector<PSimHit>(tag); }),
+      caloRechitTags_(pset.getParameter<std::vector<edm::InputTag>>("caloRecHits")),
+      caloRechitCollectionTokens_(edm::vector_transform(
+          caloRechitTags_, [this](const edm::InputTag& tag) { return mayConsume<edm::View<CaloRecHit>>(tag); })),
+      scCollectionToken_(consumes<SimClusterCollection>(pset.getParameter<edm::InputTag>("simClusters"))) {
+  for (auto& tag : caloRechitTags_) {
+    const std::string& label = tag.instance();
+    produces<edm::Association<SimClusterCollection>>(label + "ToSimClus");
+  }
+  produces<std::unordered_map<int, float>>();
```

Comment:
is this going to be persisted or just used as a transient object?
I'm not sure it's a good idea to stream out an `unordered_map`, at least long term. 

The type also does not provide any particular clarity or help to know what's mapped to what.
Can some other type be considered instead?
From the implementation it looks like the key is just the index in `SimClusterCollection`. A ValueMap is perhaps the simplest (even though it may be larger in size); how sparse is this mapping?